DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_template/src/fpga_template.cpp

```diff
@@ -0,0 +1,114 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class VectorAddID;
+
+class VectorAdd {
+ public:
+  int *a_in;
+  int *b_in;
+  int *c_out;
+  int len;
+
+  void operator()() const {
+    for (int idx = 0; idx < len; idx++) {
+      int a_val = a_in[idx];
+      int b_val = b_in[idx];
+      int sum = a_val + b_val;
+      c_out[idx] = sum;
+    }
+  }
+};
+
+#define VECT_SIZE 256
+
+int main() {
+  bool passed = false;
+
+  try {
+// This design is tested with 2023.0, but also accounts for a syntax change in
+// 2023.1
+#if __INTEL_CLANG_COMPILER >= 20230100
+#if FPGA_SIMULATOR
+  auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+  auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+  auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+#elif __INTEL_CLANG_COMPILER >= 20230000
+#if FPGA_SIMULATOR
+    auto selector = sycl::ext::intel::fpga_simulator_selector{};
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector{};
+#else  // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector{};
+#endif
+#else
+    assert(false) && "this design requires oneAPI 2023.0 or 2023.1!"
```

Comment:
The older style also works for older releases, but users with older releases such as 2022.2 are expected to fetch the 2022.2 tag of the code samples, which won't include this sample (as it uses features of the compiler that did not exist then).
Why is this sample doing this special handling for 2023.0 and 2023.1?
-> Paul wants to distribute this sample to selected customers who have the 2023.0 version of the compiler (a.k.a. not wait until the 2023.1 version of the compiler is out)
-> The 2023.1 version of the compiler needs the new selectors.

So this approach is temporary: until we get close to the 2023.1 release. Once we are a few weeks away from the release, we'll remove the logic for handling the 2023.0 case, it will then only contain the new selectors, as expected for the 2023.1 release.

Reply:
Having the #ifdef is OK to be flexible.   My point was that users don't always limit the code and toolkit combinations to what we assume they should, so being defensive and broader in what is supported is usually better.

Reply:
None of the code samples will work with a previous version of the compiler due to headers being modified etc. 
So we are not trying to get full compatibility here, just the two cases we are interested in. 
We expect the users to read the top level readme which starts by stating that they should fetch the appropriate version of the samples that matches their compiler version